### ISSUE TITLE
Rename (refactoring) works even though methods are unsaved

### ISSUE THREAD

**Issue by fuhrmanator** [guillep/pharo-core#2821]
I was modifying a method (`methodA`) and decided one of the calls to another method (`methodB`) could have a better name. So I spontaneously did rename by right-clicking on `methodB` and renamed it to `methodC`. The rename worked, but it modified (and saved) `methodA` which I was already modifying. The editor then shows the red corner to say the method was modified elsewhere. Yes, technically I allowed it, but I was unaware of the negative consequence.

The workaround (to not lose data) is to paste the still-open copy of `methodA` into another editor (playground) and cancel, then try to figure out the differences between the rename and what was already changed (or to save that copy and rename `methodB` to `methodC` manually within `methodA`).

A great solution (used in other IDEs) is simply to disallow the refactoring (e.g., the rename) if any methods (especially ones that will be affected) are unsaved. In my case, I would have happily saved the edit of `methodA` before applying the rename, or waited until I was done with the edit.

**Comment by dionisiydk** [comment]
Hi Fuhrman.

I do not understand what you do not like with current approach. You do not
loose a code in dirty editors. It still shows what you typed. You can
accept them or cancel them. So what the problem? Why you want to "list all
dirty editors" before refactoring is actually applied? To me it is kind of
modal behaviour which is always better to avoid.

Notice that you can do cmd+l to reset changes and then cmd+z to give them
back. Very simple to check what was modified in actually installed version
of method.

чт, 14 мар. 2019 г. в 13:34, C. Fuhrman <notifications@github.com>:

**Comment by fuhrmanator** [comment]
Hi Denis,

On Thu, Mar 14, 2019, 18:23 Denis Kudriashov <notifications@github.com>
wrote:

> Hi Fuhrman.
>
> I do not understand what you do not like with current approach. You do not
> loose a code in dirty editors. It still shows what you typed. You can
> accept them or cancel them.


It's true. However, the dirty editor has my changes from before the rename
(but not the rename), and the saved version has the rename (but not the
changes from before the rename). Neither version is correct. I am not a
long time user of Pharo, so I expect it to work differently.

So what the problem? Why you want to "list all
> dirty editors" before refactoring is actually applied? To me it is kind of
> modal behaviour which is always better to avoid.
>
> Notice that you can do cmd+l to reset changes and then cmd+z to give them
> back. Very simple to check what was modified in actually installed version
> of method.
>

If I do reset changes, I get the version with the renames, but lose the
changes I made before I started the rename. It doesn't seem standard to me
that ctrl-z (undo) should reapply changes I didn't accept. Again, I'm new
to Pharo and ctrl-l is like "revert" in my paradigm. I would not expect an
undo to reapply my changes.

Perhaps I need a more concrete example to help with understanding? I will
give it a try tomorrow morning on my PC.

Thanks for the reply.

Cris
